agent/bench-scripts/pbench-user-benchmark

```diff
@@ -30,209 +31,324 @@ benchmark="pbench-user-benchmark"
+		printf -- " | pbench-tool-trigger ${iteration} ${benchmark_run_dir} ${tool_group}" >> ${benchmark_run_cmd}
+	fi
+	printf -- "\n" >> ${benchmark_run_cmd}
+	chmod +x ${benchmark_run_cmd}
 
-echo "Running $benchmark_bin"
-log "[$script_name] Running $benchmark_bin"
+	if [[ ${use_tool_triggers} -eq 0 ]]; then
+		if [[ ! -z "${pbench_pre}" ]]; then
+			log "[${script_name}]: Running ${pbench_pre} for iteration ${iteration_name}"
+			eval ${pbench_pre}
+			if [[ ${?} != 0 ]]; then
+				error_log "[${script_name}]: the --pbench-pre script, '${pbench_pre}', failed"
+				exit 1
+			fi
+		fi
+		pbench-start-tools --group=${tool_group} --iteration=${iteration} --dir=${benchmark_results_dir}
+	fi
 
-SECONDS=0
-$benchmark_run_cmd
-benchmark_duration=$SECONDS
-# make it predictable in the unittest environment
-if [[ $_PBENCH_BENCH_TESTS == 1 ]]; then
-	benchmark_duration=0
-fi
+	_msg="Running ${iter_benchmark_bin} for iteration ${iteration_name}"
+	echo "${_msg}"
+	log "[${script_name}] ${_msg}"
 
-if [ $use_tool_triggers -eq 0 ]; then
-	pbench-stop-tools --group=$tool_group --iteration=$iteration --dir=$benchmark_results_dir
-	pbench-postprocess-tools --group=$tool_group --iteration=$iteration --dir=$benchmark_results_dir
-	ln -s $(basename ${benchmark_results_dir}) $(dirname ${benchmark_results_dir})/reference-result
-else
-	result_dirs=$(ls -1d ${benchmark_run_dir}/*/sample1)
-	for rdir in $result_dirs; do
-		iteration_dir=$(basename $(dirname ${rdir}))
-		pbench-postprocess-tools --group=${tool_group} --iteration=${iteration_dir} --dir=${rdir}
-		ln -s $(basename ${rdir}) $(dirname ${rdir})/reference-result
-	done
-fi
-pbench-metadata-log --group=$tool_group --dir=$benchmark_run_dir end
+	SECONDS=0
+	${benchmark_run_cmd}
+	benchmark_duration=${SECONDS}
+	# make it predictable in the unittest environment
+	if [[ ${_PBENCH_BENCH_TESTS} == 1 ]]; then
+		benchmark_duration=0
+	fi
 
-if [[ ! -z $pbench_post ]]; then
-	log "[$script_name]: Running $pbench_post"
-	eval $pbench_post
-	if [[ $? != 0 ]]; then
-		error_log "[$script_name]: the script executed using pbench_post flag failed"
-		exit 1
+	if [[ $use_tool_triggers -eq 0 ]]; then
+		pbench-stop-tools --group="${tool_group}" --iteration="${iteration}" --dir="${benchmark_results_dir}"
+		pbench-postprocess-tools --group="${tool_group}" --iteration="${iteration}" --dir="${benchmark_results_dir}"
+		if [[ ! -z ${pbench_post} ]]; then
+			log "[${script_name}]: Running ${pbench_post} for iteration ${iteration_name}"
+			eval ${pbench_post}
+			if [[ ${?} != 0 ]]; then
+				error_log "[${script_name}]: the --pbench-post script, '${pbench_post}', failed"
+				exit 1
+			fi
+		fi
+	fi
+	return ${benchmark_duration}
+}
+
+iter_num=${iteration}
+lineno=0
+total_duration=0
+declare -a parts
+# NOTE: the following "while" loop read from FD 3 for the input file so that
+# the while loop sub-shell does not share FD 0 (stdin) with the executed
+# benchmark script's FD 0. This can lead to problems when benchmarks mess
```

Comment:
"... with the executed benchmark script." 

"This can lead ..." --> "Otherwise, benchmarks messing around with FD 0 can lead to problems."

Reply:
Fixed.